go/vt/vtgate/resilient_srv_topo_server.go

```diff
@@ -280,33 +299,50 @@ func (server *ResilientSrvTopoServer) GetSrvKeyspace(ctx context.Context, cell,
 	entry.mutex.Lock()
 	defer entry.mutex.Unlock()
 
-	// If the entry is fresh enough, return it
-	if time.Now().Sub(entry.insertionTime) < server.cacheTTL {
+	// If the entry exists, return it
+	if entry.value != nil {
 		return entry.value, entry.lastError
 	}
 
-	// not in cache or too old, get the real value
-	newCtx, cancel := context.WithTimeout(context.Background(), *srvTopoTimeout)
-	defer cancel()
+	// not in cache, get the real value
+	newCtx := context.Background()
 
-	result, err := server.topoServer.GetSrvKeyspace(newCtx, cell, keyspace)
+	// start watching
+	notifications, stopWatching, err := server.topoServer.WatchSrvKeyspace(newCtx, cell, keyspace)
 	if err != nil {
-		if entry.insertionTime.IsZero() {
-			server.counts.Add(errorCategory, 1)
-			log.Errorf("GetSrvKeyspace(%v, %v, %v) failed: %v (no cached value, caching and returning error)", newCtx, cell, keyspace, err)
-		} else {
-			server.counts.Add(cachedCategory, 1)
-			log.Warningf("GetSrvKeyspace(%v, %v, %v) failed: %v (returning cached value: %v %v)", newCtx, cell, keyspace, err, entry.value, entry.lastError)
-			return entry.value, entry.lastError
-		}
+		entry.lastError = err
+		entry.lastErrorCtx = newCtx
```

Comment:
This should be ctx, the point is for the web page to show who originated the query, otherwise this newCtx is empty.